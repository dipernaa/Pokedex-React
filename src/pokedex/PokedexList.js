import React, { Component } from 'react';
import { browserHistory } from 'react-router';
import { List, ListItem } from 'material-ui/List';
import TextField from 'material-ui/TextField';
import { routeNames } from '../routes';

const styles = {
  container: {
    display: 'flex',
    flexDirection: 'column'
  },
  search: {
    marginLeft: 'auto',
    marginRight: 'auto'
  }
};

class PokedexList extends Component {

  constructor(props) {
    super(props);

    this.state = {
      filteredPokedex: props.pokedex
    };
  }

  filterPokedex = (value) => {
    if (value) {
      this.setState({
        filteredPokedex: this.props.pokedex.filter(item => item.name.indexOf(value) >= 0)
      });
    } else {
      this.setState({
        filteredPokedex: this.props.pokedex
      });
    }
  }

  render() {
    if (!this.state.filteredPokedex) {
      return null;
    }

    return (
      <div style={styles.container}>
        <TextField
          hintText="Search"
          onChange={(event, value) => this.filterPokedex(value)}
          style={styles.search}
        />
        <br/>

        <div>
          <List>
            {this.state.filteredPokedex.map((pokemon, index) => (
              <ListItem
                key={index}
                className="capitalize"
                onClick={() => browserHistory.push(`${routeNames.detail}?url=${pokemon.url}`)}
                primaryText={pokemon.name}
              />
            ))}
          </List>
        </div>
      </div>
    );
  }
}

export default PokedexList;
