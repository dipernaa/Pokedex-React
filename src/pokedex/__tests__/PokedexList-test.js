import React from 'react';
import { shallow } from 'enzyme';
import { browserHistory } from 'react-router';
import toJson from 'enzyme-to-json';
import PokedexList from '../PokedexList';

jest.mock('react-router', () => ({
  browserHistory: {
    push: jest.fn()
  }
}));

const testProps = {
  pokedex: [
    {
      name: 'pokedude',
      url: 'http://www.some.url'
    },
    {
      name: 'pokeman',
      url: 'http://www.worm.url'
    },
    {
      name: 'pokelizard',
      url: 'http://www.squad.url'
    }
  ]
};

it('renders null when pokemon is not present', () => {
  const wrapper = shallow(
    <PokedexList/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('renders a PokedexList when pokemon is present', () => {
  const wrapper = shallow(
    <PokedexList {...testProps}/>
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('resets pokedex when search field changes without a value', () => {
  const wrapper = shallow(
    <PokedexList {...testProps}/>
  );

  const searchField = wrapper.find('TextField');
  searchField.props().onChange({}, '');

  const filteredPokemon = wrapper.state('filteredPokedex');
  expect(filteredPokemon.length).toEqual(testProps.pokedex.length);
});

it('filters pokedex when search field changes with value', () => {
  const wrapper = shallow(
    <PokedexList {...testProps}/>
  );

  const searchField = wrapper.find('TextField');
  searchField.props().onChange({}, 'dude');

  const filteredPokemon = wrapper.state('filteredPokedex');
  expect(filteredPokemon.length).toEqual(1);
  expect(filteredPokemon[0].name).toEqual(testProps.pokedex[0].name);
  expect(filteredPokemon[0].url).toEqual(testProps.pokedex[0].url);
});

it('calls browserHistory.push() when list item is press', () => {
  const wrapper = shallow(
    <PokedexList {...testProps}/>
  );

  const listItem = wrapper.find('ListItem').first();
  listItem.props().onClick();

  expect(browserHistory.push.mock.calls.length).toBe(1);
});
