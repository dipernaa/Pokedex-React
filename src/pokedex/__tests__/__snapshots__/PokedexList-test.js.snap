// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a PokedexList when pokemon is present 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <TextField
    disabled={false}
    floatingLabelFixed={false}
    fullWidth={false}
    hintText="Search"
    multiLine={false}
    onChange={[Function]}
    rows={1}
    style={
      Object {
        "marginLeft": "auto",
        "marginRight": "auto",
      }
    }
    type="text"
    underlineShow={true}
  />
  <br />
  <div>
    <List>
      <ListItem
        autoGenerateNestedIndicator={true}
        className="capitalize"
        containerElement="span"
        disableKeyboardFocus={false}
        disabled={false}
        initiallyOpen={false}
        insetChildren={false}
        nestedItems={Array []}
        nestedLevel={0}
        onClick={[Function]}
        onKeyboardFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onNestedListToggle={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        open={null}
        primaryText="pokedude"
        primaryTogglesNestedList={false}
        secondaryTextLines={1}
      />
      <ListItem
        autoGenerateNestedIndicator={true}
        className="capitalize"
        containerElement="span"
        disableKeyboardFocus={false}
        disabled={false}
        initiallyOpen={false}
        insetChildren={false}
        nestedItems={Array []}
        nestedLevel={0}
        onClick={[Function]}
        onKeyboardFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onNestedListToggle={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        open={null}
        primaryText="pokeman"
        primaryTogglesNestedList={false}
        secondaryTextLines={1}
      />
      <ListItem
        autoGenerateNestedIndicator={true}
        className="capitalize"
        containerElement="span"
        disableKeyboardFocus={false}
        disabled={false}
        initiallyOpen={false}
        insetChildren={false}
        nestedItems={Array []}
        nestedLevel={0}
        onClick={[Function]}
        onKeyboardFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onNestedListToggle={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        open={null}
        primaryText="pokelizard"
        primaryTogglesNestedList={false}
        secondaryTextLines={1}
      />
    </List>
  </div>
</div>
`;

exports[`renders null when pokemon is not present 1`] = `null`;
